### PR TITLE
Tofu keystore

### PR DESCRIPTION
**Pull request by ianopolous**
This implements a TOFU keystore which is stored in the user's (Peergos) root directory as a hidden file. This largely decentralizes the PKI part of the corenode in terms of traffic.

### CODE REVIEW

**Comment by ianopolous** [comment]
I think this is good to go now.

File: src/peergos/shared/user/EntryPoint.java [hunk 1]
```diff
@@ -36,6 +37,21 @@ public EntryPoint(FilePointer pointer, String owner, Set<String> readers, Set<St
         return ArrayOps.concat(nonce, key.encrypt(serialize(), nonce));
     }
 
+    public CompletableFuture<Boolean> isValid(String path, NetworkAccess network) {
```
Comment (cboddy) [inline src/peergos/shared/user/EntryPoint.java]: docstring please

File: src/peergos/shared/user/TofuKeyStore.java [hunk 2]
```diff
@@ -22,19 +24,33 @@ public TofuKeyStore(Map<String, List<UserPublicKeyLink>> chains) {
         return Optional.of(chain.get(chain.size() - 1).owner);
     }
 
+    public Optional<String> getUsername(PublicSigningKey signer) {
+        String name = reverseLookup.get(signer);
+        if (name == null)
```
Comment (cboddy) [inline src/peergos/shared/user/TofuKeyStore.java]: ``return  Optional.ofNullable(name);``

File: src/peergos/shared/corenode/UserPublicKeyLink.java [hunk 3]
```diff
@@ -136,7 +136,7 @@ public static UsernameClaim fromCbor(CborObject cbor) {
                 throw new IllegalStateException("Invalid cbor for Username claim: " + cbor);
             String username = ((CborObject.CborString)((CborObject.CborList) cbor).value.get(0)).value;
             LocalDate expiry = LocalDate.parse(((CborObject.CborString)((CborObject.CborList) cbor).value.get(1)).value);
-            byte[] signedContents = ((CborObject.CborByteArray)((CborObject.CborList) cbor).value.get(0)).value;
+            byte[] signedContents = ((CborObject.CborByteArray)((CborObject.CborList) cbor).value.get(2)).value;
```
Comment (cboddy) [inline src/peergos/shared/corenode/UserPublicKeyLink.java]: 2????

**Pull request merged**